File: kotlin/app/src/commonMain/kotlin/kotlin_test/CollectionBenchmark.kt
package kotlin_test

import kotlin.random.Random

object CollectionBenchmark {

    private const val times = 10000
    private const val timesShort = 100
    private const val listSize = 10000
    private var timer = 0L

    fun add10k(): Double {
        timer = Util.getNanoTime()
        add(times)
        timer = Util.getNanoTime() - timer

        return timer / times.toDouble()
    }

    fun read10k(): Double {
        val list = generateList(times)
        val dummy: Int

        timer = Util.getNanoTime()
        dummy = read(list)
        timer = Util.getNanoTime() - timer

        println(dummy)
        return timer / times.toDouble()
    }

    fun read10PercentRandom(): Double {
        val list = generateList(times * 10)
        val toRead = generateRandomIndexes(times, times * 10)
        val dummy: Int

        timer = Util.getNanoTime()
        dummy = readRandom(list, toRead)
        timer = Util.getNanoTime() - timer

        println(dummy)
        return timer / times.toDouble()
    }

    fun remove10PercentRandom(): Double {
        val list = generateList(times * 10)
        val toRemove = generateRandomIndexes(times, times * 10, true)

        timer = Util.getNanoTime()
        removeRandom(list, toRemove)
        timer = Util.getNanoTime() - timer

        return timer / times.toDouble()
    }

    fun filter(): Double {
        var result = 0L

        val minIndex = listSize / 2
        var list: ArrayList<TestObject>
        repeat(timesShort) {
            list = generateList()
            timer = Util.getNanoTime()
            filter(list, minIndex)
            result += Util.getNanoTime() - timer
        }

        return result / timesShort.toDouble()
    }

    fun sort(): Double {
        var result = 0L

        val list = generateList()
        repeat(timesShort) {
            list.shuffle()
            timer = Util.getNanoTime()
            sort(list)
            result += Util.getNanoTime() - timer
        }

        return result / timesShort.toDouble()
    }

    private fun add(times: Int = 10000): Int {
        val list = ArrayList<TestObject>()
        repeat(times) {
            list.add(TestObject(it, "item$it"))
        }
        return list.size
    }

    private fun read(list: List<TestObject>): Int {
        var dummy = 0
        for (item in list) {
            dummy += item.index
        }
        return dummy
    }

    private fun readRandom(list: List<TestObject>, toRead: List<Int>): Int {
        var dummy = 0
        for (i in toRead) {
            dummy += list[i].index
        }
        return dummy
    }

    private fun removeRandom(list: ArrayList<TestObject>, toRemove: List<Int>): Int {
        for (i in toRemove) {
            list.removeAt(i)
        }
        return list.size
    }

    private fun filter(list: ArrayList<TestObject>, minIndex: Int): Int {
        list.filter { it.index > minIndex }
        return list.last().index
    }

    private fun sort(list: ArrayList<TestObject>): Int {
        list.sortBy { it.index }
        return list.last().index
    }

    private fun generateList(size: Int = listSize): ArrayList<TestObject> {
        val list = ArrayList<TestObject>()
        repeat(size) {
            list.add(TestObject(it, "item$it"))
        }
        return list
    }

    private fun generateRandomIndexes(size: Int, until: Int, reduce: Boolean = false): List<Int> {
        val list = ArrayList<Int>()
        repeat(size) {
            list.add(Random.nextInt(until - (if (reduce) it else 0)))
        }
        return list
    }

}

File: flutter/lib/CollectionBenchmark.dart
import 'dart:math';

import 'helper/TestObject.dart';

class CollectionBenchmark {
  static const int _times = 10000;
  static const int _timesShort = 100;
  static const int _listSize = 10000;
  int _timer = 0;

  Future<double> add10k() async {
    _timer = DateTime.now().microsecondsSinceEpoch;
    _add(_times);
    _timer = DateTime.now().microsecondsSinceEpoch - _timer;

    return _timer / _times;
  }

  Future<double> readAll() async {
    List<TestObject> list = _generateList(_times);
    int dummy;

    _timer = DateTime.now().microsecondsSinceEpoch;
    dummy = _read(list);
    _timer = DateTime.now().microsecondsSinceEpoch - _timer;

    print(dummy);
    return _timer / _times;
  }

  Future<double> read10PercentRandom() async {
    List<TestObject> list = _generateList(_times * 10);
    List<int> toRead = _generateRandomIndexes(_times, _times * 10, false);
    int dummy;

    _timer = DateTime.now().microsecondsSinceEpoch;
    dummy = _readRandom(list, toRead);
    _timer = DateTime.now().microsecondsSinceEpoch - _timer;

    print(dummy);
    return _timer / _times;
  }

  Future<double> remove10PercentRandom() async {
    List<TestObject> list = _generateList(_times * 10);
    List<int> toRemove = _generateRandomIndexes(_times, _times * 10, true);

    _timer = DateTime.now().microsecondsSinceEpoch;
    _removeRandom(list, toRemove);
    _timer = DateTime.now().microsecondsSinceEpoch - _timer;

    return _timer / _times;
  }

  Future<double> filter() async {
    int result = 0;

    int minIndex = _listSize ~/ 2;
    List<TestObject> list;
    for (int i = 0; i < _timesShort; i++) {
      list = _generateList(_listSize);
      _timer = DateTime.now().microsecondsSinceEpoch;
      _filter(list, minIndex);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    return result / _timesShort;
  }

  Future<double> sort() async {
    int result = 0;

    List<TestObject> list = _generateList(_listSize);
    for (int i = 0; i < _timesShort; i++) {
      list.shuffle();
      _timer = DateTime.now().microsecondsSinceEpoch;
      _sort(list);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }
    return result / _timesShort;
  }

  int _add(int times) {
    List<TestObject> list = List();
    for(int i = 0; i < times; i++) {
      list.add(TestObject(i, 'item $i'));
    }
    return list.length;
  }

  int _read(List<TestObject> list) {
    int dummy = 0;
    for (TestObject i in list) {
      dummy += i.index;
    }
    return dummy;
  }

  int _readRandom(List<TestObject> list, List<int> toRead) {
    int dummy = 0;
    for (int i in toRead) {
      dummy += list[i].index;
    }
    return dummy;
  }

  int _removeRandom(List<TestObject> list, List<int> toRemove) {
    for (int i in toRemove) {
      list.removeAt(i);
    }
    return list.length;
  }

  int _filter(List<TestObject> list, int minIndex) {
    list.removeWhere((i) => i.index > minIndex);
    return list.length;
  }

  int _sort(List<TestObject> list) {
    list.sort((a, b) => a.index.compareTo(b.index));
    return list.last.index;
  }

  List<TestObject> _generateList(int size) {
    return List.generate(size, (i) => new TestObject(i, 'item $i'));
  }

  List<int> _generateRandomIndexes(int size, int maxIndex, bool reduce) {
    return List.generate(size, (i) => Random().nextInt(maxIndex - (reduce ? i : 0)));
  }

}